File: Medium/328 - OddEvenLinkedList/Solution.cs
// Question 328. Odd Even Linked List (https://leetcode-cn.com/problems/odd-even-linked-list/)

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class ListNode 
{
    public int val;
    public ListNode next;
    public ListNode(int val=0, ListNode next=null) 
    {
        this.val = val;
        this.next = next;
    }
}

/// <summary>
/// 执行用时：108 ms, 在所有 C# 提交中击败了 84.71% 的用户
/// 内存消耗：25.7 MB, 在所有 C# 提交中击败了 9.41% 的用户
/// </summary>
public class Solution 
{
    public ListNode OddEvenList(ListNode head) 
    {
        if (head == null || head.next == null) { return head; }
        ListNode HEAD_POINTER = head;
        ListNode EVEN_HEAD_POINTER = head.next;
        ListNode evenNode = head.next;

        while (evenNode != null && evenNode.next != null)
        {
            head.next = evenNode.next;
            head = head.next;

            evenNode.next = head.next;
            evenNode = evenNode.next;
        }

        head.next = EVEN_HEAD_POINTER;
        return HEAD_POINTER;
    }
}

File: Medium/328 - OddEvenLinkedList/solution.ts
// Question 328. Odd Even Linked List (https://leetcode-cn.com/problems/odd-even-linked-list/)

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

/**
 * 执行用时：104 ms, 在所有 typescript 提交中击败了 41.67% 的用户
 * 内存消耗：41.6 MB, 在所有 typescript 提交中击败了 16.67% 的用户
 *
 * @param {(ListNode | null)} head
 * @returns {(ListNode | null)}
 */
function oddEvenList(head: ListNode | null): ListNode | null {
  if (!head || !head.next) {
    return head;
  }

  const HEAD_POINTER: ListNode = head;
  const EVEN_HEAD_POINTER: ListNode = head.next;
  let evenPointer: ListNode | null = EVEN_HEAD_POINTER;

  while (evenPointer && evenPointer.next) {
    head.next = evenPointer.next;
    head = head.next;

    evenPointer.next = head.next;
    evenPointer = evenPointer.next;
  }

  head.next = EVEN_HEAD_POINTER;

  return HEAD_POINTER;
}
